pkg/api/v1/summary.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleCloudPlatform/testgrid/config"
+	apipb "github.com/GoogleCloudPlatform/testgrid/pb/api/v1"
+	summarypb "github.com/GoogleCloudPlatform/testgrid/pb/summary"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer"
+)
+
+const (
+	// PASSING is "PASSING"
+	PASSING = "PASSING"
+	// FAILING is "FAILING"
+	FAILING = "FAILING"
+	// FLAKY is "FLAKY"
+	FLAKY = "FLAKY"
+	// ACCEPTABLE is "ACCEPTABLE"
+	ACCEPTABLE = "ACCEPTABLE"
+	// STALE is "STALE"
+	STALE = "STALE"
+	// BROKEN is "BROKEN"
+	BROKEN = "BROKEN"
+	// UNKNOWN is "UNKNOWN"
+	UNKNOWN = "UNKNOWN"
+	// PENDING is "PENDING"
+	PENDING = "PENDING"
+)
+
+var (
+	// These should be in line with the TabStatus enum in state.proto.
+	tabStatusStr = map[summarypb.DashboardTabSummary_TabStatus]string{
+		summarypb.DashboardTabSummary_PASS:       PASSING,
+		summarypb.DashboardTabSummary_FAIL:       FAILING,
+		summarypb.DashboardTabSummary_FLAKY:      FLAKY,
+		summarypb.DashboardTabSummary_STALE:      STALE,
+		summarypb.DashboardTabSummary_BROKEN:     BROKEN,
+		summarypb.DashboardTabSummary_PENDING:    PENDING,
+		summarypb.DashboardTabSummary_ACCEPTABLE: ACCEPTABLE,
+	}
+)
+
+// fetchSummary returns the summary struct as defined in summary.proto.
+// Returns an error iff the scope refers to non-existent bucket OR server fails to read the summary.
+func (s *Server) fetchSummary(ctx context.Context, scope, dashboard string) (*summarypb.DashboardSummary, error) {
+	configPath, _, err := s.configPath(scope)
+
+	summaryPath, err := summarizer.SummaryPath(*configPath, s.SummaryPathPrefix, dashboard)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the summary path: %v", err)
+	}
+
+	summary, _, _, err := summarizer.ReadSummary(ctx, s.Client, *summaryPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download summary at %v: %v", summaryPath.String(), err)
+	}
+
+	return summary, nil
+}
+
+// ListTabSummaries returns the list of tab summaries for the particular dashboard.
+// Returns an error iff dashboard does not exist OR the server can't read summaries from GCS bucket.
+func (s *Server) ListTabSummaries(ctx context.Context, req *apipb.ListTabSummariesRequest) (*apipb.ListTabSummariesResponse, error) {
+	ctx, cancel := context.WithTimeout(ctx, s.Timeout)
+	defer cancel()
+
+	scope := req.GetScope()
+	cfg, err := s.getConfig(ctx, logrus.WithContext(ctx), scope)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch config from {%q}: %v", scope, err)
+	}
+
+	cfg.Mutex.RLock()
+	defer cfg.Mutex.RUnlock()
+
+	dashboardKey := config.Normalize(req.GetDashboard())
+	if _, ok := cfg.NormalDashboard[dashboardKey]; !ok {
+		return nil, fmt.Errorf("dashboard {%q} not found", dashboardKey)
+	}
+
+	summary, err := s.fetchSummary(ctx, scope, dashboardKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch summary for dashboard {%q}: %v", dashboardKey, err)
+	}
+
+	// TODO(sultan-duisenbay): nil summary?
```

Comment:
Not clear from this TODO what issue you're talking about here. Do you mean just checking if summary is nil AND error is nil?

Reply:
Yes, I added the check now.